### ISSUE TITLE
no-use-before-define: Incorrectly warns when the 1st of two function statements references the 2nd.

### ISSUE THREAD

**Issue by iskitz** [eslint/eslint#1485]
| ... | ... |
| --- | --- |
| **version** | 0.9.2 |
| **rule** | [no-use-before-define](http://eslint.org/docs/rules/no-use-before-define.html) |
| **code** | See below |
| **expected** | No warning |
| **observed** | `[no-use-before-define] b was used before it was defined` |

ESLint throws the warning above when linting this code

``` javascript
function a () { b(); }
function b () { /* do something other than calling function a.*/ }
```

This shouldn't be a warning because [function declarations are hoisted](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function#Function_declaration_hoisting) which means `function b()` will always be available to `function a()` and vice versa.

This is confirmed via the following code which tests for and outputs that `Both a & b exist: true`:

``` javascript
console.log ('Both a & b exist: ' + !!(a && b));

function a () { b(); }
function b () { /* do something other than calling function a.*/ }
```

Even if `function a()` was an [immediately-invoked function expression](https://en.wikipedia.org/wiki/Immediately-invoked_function_expression) it would still be able to access `function b()` for the same reason.
